object/datastore.go

```diff
@@ -167,43 +165,67 @@ func (d Datastore) useServiceTicketHostName(name string) bool {
 	return false
 }
 
+type datastoreServiceTicketHost struct{}
+
+var datastoreServiceTicketHostKey = datastoreServiceTicketHost{}
+
+// HostContext returns a Context where the given host will be used for datastore HTTP access
+// via the ServiceTicket method.
+func (d Datastore) HostContext(ctx context.Context, host *HostSystem) context.Context {
+	return context.WithValue(ctx, datastoreServiceTicketHostKey, host)
+}
+
 // ServiceTicket obtains a ticket via AcquireGenericServiceTicket and returns it an http.Cookie with the url.URL
-// that can be used along with the ticket cookie to access the given path.
+// that can be used along with the ticket cookie to access the given path.  An host is chosen at random unless the
+// the given Context was created with a specific host via the HostContext method.
 func (d Datastore) ServiceTicket(ctx context.Context, path string, method string) (*url.URL, *http.Cookie, error) {
-	// We are uploading to an ESX host
-	u := &url.URL{
-		Scheme: d.c.URL().Scheme,
-		Host:   d.c.URL().Host,
-		Path:   fmt.Sprintf("/folder/%s", path),
-		RawQuery: url.Values{
-			"dsName": []string{d.Name()},
-		}.Encode(),
-	}
+	u := d.NewURL(path)
 
-	if !d.useServiceTicket() {
-		return u, nil, nil
-	}
+	var host *HostSystem
+
+	h := ctx.Value(datastoreServiceTicketHostKey)
```

Comment:
you could just cast h in place without checking for nil OK: https://play.golang.org/p/T6cKj0hHM3